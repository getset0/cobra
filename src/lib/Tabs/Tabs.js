import React from "react";
import PropTypes from "prop-types";

class Tabs extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      selectedItem: this.props.selectedItem || 0
    };
  }
  handleItemClick = index => {
    this.setState(
      { selectedItem: index },
      () => this.props.onStateChange && this.props.onStateChange(index)
    );
  };
  render() {
    return this.props.children({
      selectedItem: this.state.selectedItem,
      handleItemClick: this.handleItemClick
    });
  }
}

Tabs.propTypes = {
  children: PropTypes.func.isRequired
};

export default Tabs;
