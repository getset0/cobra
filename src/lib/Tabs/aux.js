import styled, { css } from "styled-components";
import PropTypes from "prop-types";
import { theme } from "../constants";

export const TabButtons = styled.div`
  display: flex;
  padding: 5px;
  justify-content: space-around;
  text-transform: uppercase;
`;

TabButtons.propTypes = {
  children: PropTypes.array.isRequired
};

export const TabButton = styled.div`
  cursor: pointer;
  padding: 10px 3px 6px 3px;
  text-align: center;
  border-bottom: 2px solid transparent
  transition: all .3s ease-out;
  color: ${props => props.theme.lightBlueColor};
  width: ${props => props.width}%
  &:hover {
    opacity: 0.75;
    color: ${props => props.theme.primaryColor}
  }

  &:active {
    opacity: 1;
    color: ${props => props.theme.primaryColor}
  }

  ${props => props.isSelected && css`
    opacity: 1;
    border-bottom: 2px solid ${props => props.theme.primaryColor}
    color: ${props => props.theme.primaryColor}
  `}
`;

TabButton.propTypes = {
  theme: PropTypes.object,
  width: PropTypes.number.isRequired,
  isSelected: PropTypes.bool.isRequired,
  onClick: PropTypes.func.isRequired
};

TabButton.defaultProps = {
  theme: theme
};

export const TabItems = styled.div`
  position: relative;
  min-height: 160px;
`;

TabButtons.propTypes = {
  children: PropTypes.array.isRequired
};

export const TabItem = styled.div`
  position: absolute;
  transition: all .3s ease-out;
  height: 100%;
  width: 100%;
  ${props => props.isSelected ? css`
    opacity: 1;
  ` : css`
    opacity: 0;
  `}
`;

TabButton.propTypes = {
  isSelected: PropTypes.bool.isRequired
};
