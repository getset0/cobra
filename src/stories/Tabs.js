import React from "react";
import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { withBackgrounds } from "@storybook/addon-backgrounds";
import { backgroundOptions } from "./util";
import { Tabs, TabButtons, TabButton, TabItem, TabItems } from "../lib";

storiesOf("Tabs", module)
  .addDecorator(withInfo)
  .addDecorator(withBackgrounds(backgroundOptions))
  .add("Tabs", () => (
    <Tabs>
      {({ selectedItem, handleItemClick }) => (
        <React.Fragment>
          <TabButtons>
            {testTabItems.map((item, index) => (
              <TabButton
                width={100 / testTabItems.length}
                key={item.title}
                isSelected={selectedItem === index}
                onClick={() => handleItemClick(index)}
              >
                {item.title}
              </TabButton>
            ))}
          </TabButtons>
          <TabItems>
            {testTabItems.map((item, index) => (
              <TabItem key={index} isSelected={selectedItem === index}>
                {item.contents}
              </TabItem>
            ))}
          </TabItems>
        </React.Fragment>
      )}
    </Tabs>
  ));

const testTabItems = [
  {
    title: "Tab1 ",
    contents: (
      <div style={{ backgroundColor: "red" }}>
        <h1>Tab 1</h1>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
        <p>BC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC ABC</p>
      </div>
    )
  },
  {
    title: "Tab 2",
    contents: (
      <div style={{ backgroundColor: "blue" }}>
            DEF ABC DEF DEF DEF DEF DEF DEF DEF DEFDEFDEFDEFDEFDEFDEFDEFDEF
      </div>
    )
  },
  {
    title: "Tab 3",
    contents: (
      <div style={{ backgroundColor: "yellow" }}>
            GHIGHIGHIGHIGHIGHIGHIGHIGHIGHI GHIGHIGHIGHIGHIGHIGHIGHIGHIGHI
            GHIGHIGHIGHIGHIGHIGHIGHIGHIGHI GHIGHIGHIGHIGHIGHIGHIGHIGHIGHI
      </div>
    )
  }
];

