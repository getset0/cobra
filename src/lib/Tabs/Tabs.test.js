import React from "react";
import Tabs from "./Tabs";
import {
  TabButtons,
  TabItems,
  TabItem,
  TabButton
} from "./aux";
import "jest-styled-components";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const testTabItems = [
  {
    id: "test",
    title: "Test",
    content: <div>test</div>
  },
  {
    id: "test2",
    title: "Test2",
    content: <div>test2</div>
  }
];

const BasicTabs = (props) =>
  <Tabs {...props}>
    {({ selectedItem, handleItemClick }) => (
      <React.Fragment>
        <TabButtons>
          {testTabItems.map((item, index) => (
            <TabButton
              width={100/testTabItems.length}
              key={item.title}
              id={item.id}
              isSelected={selectedItem === index}
              onClick={() => handleItemClick(index)}
            >
              {item.title}
            </TabButton>
          ))}
        </TabButtons>
        <TabItems>
          {testTabItems.map((item, index) => (
            <TabItem
              key={index}
              isSelected={selectedItem === index}
            >
              {item.contents}
            </TabItem>
          ))}
        </TabItems>
      </React.Fragment>
    )}
  </Tabs>;

it("renders without crash", () => {
  shallow(BasicTabs());
});

it("renders Tabs correctly", () => {
  const wrapper = shallow(BasicTabs());
  const tree = toJson(wrapper);
  expect(tree).toMatchSnapshot();
});

it("test Tabs click", () => {
  const mockStateChange = jest.fn(x => x);
  const wrapper = shallow(
    BasicTabs({ onStateChange: mockStateChange })
  );

  // has 2 TabButton Component
  const buttons = wrapper.find(TabButton);
  expect(buttons).toHaveLength(2);

  const firstButton = wrapper.find("#test");
  const secondButton = wrapper.find("#test2");

  expect(firstButton).toHaveLength(1);
  expect(secondButton).toHaveLength(1);

  // click 3 times on trigger
  firstButton.prop("onClick")();
  secondButton.prop("onClick")();
  firstButton.prop("onClick")();

  // expect to have called mock 3 times
  expect(mockStateChange.mock.calls).toHaveLength(3);

  // expect mock params to be consistent
  expect(mockStateChange.mock.calls[0][0]).toBe(0);
  expect(mockStateChange.mock.calls[1][0]).toBe(1);
  expect(mockStateChange.mock.calls[2][0]).toBe(0);
});
