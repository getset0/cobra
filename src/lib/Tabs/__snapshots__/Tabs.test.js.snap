// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Tabs correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div
      id="test"
      isSelected={true}
      key="Test"
      onClick={[Function]}
      theme={
        Object {
          "blackColor": "#000000",
          "darkBlueColor": "#02469B",
          "darkGrayColor": "#4A4A4A",
          "darkestBlueColor": "#02254D",
          "darkestGrayColor": "#303030",
          "greenColor": "#76B041",
          "lightBlueColor": "#0190D0",
          "lightGrayColor": "#D8D8D8",
          "lightestBlueColor": "#00FEFE",
          "mediumGrayColor": "#979797",
          "primaryColor": "#1676B7",
          "secondaryColor": "#17BEBB",
          "whiteBackgroundColor": "#fdfdfd",
          "whiteColor": "#ffffff",
          "yellowColor": "#FFC914",
        }
      }
      width={50}
    >
      Test
    </styled.div>
    <styled.div
      id="test2"
      isSelected={false}
      key="Test2"
      onClick={[Function]}
      theme={
        Object {
          "blackColor": "#000000",
          "darkBlueColor": "#02469B",
          "darkGrayColor": "#4A4A4A",
          "darkestBlueColor": "#02254D",
          "darkestGrayColor": "#303030",
          "greenColor": "#76B041",
          "lightBlueColor": "#0190D0",
          "lightGrayColor": "#D8D8D8",
          "lightestBlueColor": "#00FEFE",
          "mediumGrayColor": "#979797",
          "primaryColor": "#1676B7",
          "secondaryColor": "#17BEBB",
          "whiteBackgroundColor": "#fdfdfd",
          "whiteColor": "#ffffff",
          "yellowColor": "#FFC914",
        }
      }
      width={50}
    >
      Test2
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div
      isSelected={true}
      key="0"
    />
    <styled.div
      isSelected={false}
      key="1"
    />
  </styled.div>
</Fragment>
`;
